### PR TITLE
[POC] Pass request context to backend methods

### PR DESCRIPTION
**Pull request by saw-jan**
POC for https://github.com/rclone/rclone/issues/7645

### CODE REVIEW

**Comment by ncw** [comment]
Here is an idea for you - rather than passing `http.Request` to every method it might be better done by adding `ctx context.Context` to each method (which it should have already really) and adding the request to the context.

**Comment by saw-jan** [comment]
> Here is an idea for you - rather than passing `http.Request` to every method it might be better done by adding `ctx context.Context` to each method (which it should have already really) and adding the request to the context.

Thanks, I will give it a shot

File: gofakes3.go [hunk 1]
```diff
@@ -184,7 +184,7 @@ func (g *GoFakeS3) httpError(w http.ResponseWriter, r *http.Request, err error)
 }
 
 func (g *GoFakeS3) listBuckets(w http.ResponseWriter, r *http.Request) error {
-	buckets, err := g.storage.ListBuckets()
+	buckets, err := g.storage.ListBuckets(r.Context())
```
Comment (saw-jan) [inline gofakes3.go:187]: @ncw I refactored it to use the context. Could you have a look again? Then I can move on to fixing tests and other usage

Comment (saw-jan) [reply gofakes3.go:187]: UPDATE: tests are fixed

**Review by ncw — APPROVED** [review]
This looks great thank you.

I'm going to merge it now. I'm slightly concerned that things will go wrong as we moved the repo to the rclone organization but we'll see what happens!

**Comment by ncw** [comment]
@saw-jan I've now moved rclone over to using the fork in its new place and adjusted the calling parameters to add the new ctx parameters so its all ready for the next part.

**Comment by saw-jan** [comment]
> @saw-jan I've now moved rclone over to using the fork in its new place and adjusted the calling parameters to add the new ctx parameters so its all ready for the next part.

Thank you. I will update the rclone PR

**Pull request merged**